### DISCUSSION TITLE
Concurrent, conflict-free writes to DuckDB database in READ_WRITE mode fails

### DISCUSSION THREAD

**Discussion by unidesigner** [duckdb/duckdb#4899] [Q&A]
I have a DuckDB file with multiple tables and I'd like to write/update individual tables concurrently. Each process only updates a single table, so there will be no conflicts or race-conditions. If I try to do this in Python with `read_only=False` I get a runtime error `RuntimeError: IO Error: Could not set lock on file`. Is there a way to disable the single-threaded write-model?

Previously, I separated out each table in its own duckdb file, but I also need to later on to do read-only queries which join across tables. As far as I read the docs, there is no way to query from multiple duckdb db files similar to what is possible with parquet files, by giving a path. Maybe this could be a useful feature to add?

**Comment by Mytherin** [comment]
Concurrent writes to the same DuckDB file from multiple processes are not supported and are unlikely to be supported in the future. Even if the processes write to disjoint tables, there will still be contention in the storage as the processes will want to write to the same WAL file, the same blocks, etc. This requires coordination between the different processes that is out-of-scope for an in-process database.

You can write to the same DuckDB file from the same process using separate connections through the use of cursors, e.g.:

```py
import duckdb
db = duckdb.connect()
con1 = db.cursor()
con2 = db.cursor()
con1.execute('create table tbl(i integer)')
con2.execute('create table tbl2(i integer)')
```

Alternatively, you can have a process open a connection to the database, write to the file, then close the database connection again, followed by the other process doing the same. This is how SQLite operates internally. DuckDB is not specifically optimized for this use case, however, so opening and closing a connection for every write is not optimized. Doing so for batch writes might have acceptable performance.

> Previously, I separated out each table in its own duckdb file, but I also need to later on to do read-only queries which join across tables. As far as I read the docs, there is no way to query from multiple duckdb db files similar to what is possible with parquet files, by giving a path. Maybe this could be a useful feature to add?
 
Absolutely. This feature is on our road-map and something we have been wanting to add for a while (see #1985).

**Reply by unidesigner** [reply]
Thank you @Mytherin for your quick and detailed reply. Well appreciated.

Your alternative suggestion sounds like something I'd like to try, as I have separate processes which take different amounts of time and have a large single write/update at the end. Sometimes, updates may be concurrent and thus one processes may not be able to acquire the lock. But I could just wait for a set amount of time (e.g. 10 seconds) and try again in a while loop until the lock can get acquired. This simple scheme for synchronization should work for my case. Do you think that is a good way forward?

Great to know about plans to implement #1985. If it happens, I may revert back to create single database files for single tables to enable concurrent writes.
